internal/controllers/core/togglebutton/reconciler.go

```diff
@@ -0,0 +1,291 @@
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, err
+}
+
+func (r *Reconciler) processClick(ctx context.Context, tb v1alpha1.ToggleButton) error {
+	uiButton := v1alpha1.UIButton{}
+	err := r.ctrlClient.Get(ctx, types.NamespacedName{Name: uibuttonName(tb.Name)}, &uiButton)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		} else {
+			return err
+		}
+	}
+
+	// if there's a new click, pass the new value through to the ConfigMap
+	if uiButton.Status.LastClickedAt.After(r.lastClickProcessTimes[tb.Name]) {
+		foundInput := false
+		var newOnValue bool
+		for _, input := range uiButton.Status.Inputs {
+			if input.Name == actionUIInputName {
+				if input.Hidden == nil {
+					return fmt.Errorf("button %q input %q was not of type 'Hidden'", uiButton.Name, input.Name)
+				}
+				switch input.Hidden.Value {
+				case turnOnInputValue:
+					newOnValue = true
+				case turnOffInputValue:
+					newOnValue = false
+				default:
+					return fmt.Errorf("button %q input %q had unexpected value %q", uiButton.Name, input.Name, input.Hidden.Value)
+				}
+				foundInput = true
+				break
+			}
+		}
+
+		if !foundInput {
+			return fmt.Errorf("UIButton %q does not have an input named %q", uiButton.Name, actionUIInputName)
+		}
+
+		ss := tb.Spec.StateSource.ConfigMap
+		if ss == nil {
+			return fmt.Errorf("ToggleButton %q has nil Spec.StateSource.ConfigMap", tb.Name)
+		}
+		var cm v1alpha1.ConfigMap
+		err := r.ctrlClient.Get(ctx, types.NamespacedName{Name: ss.Name}, &cm)
+		if err != nil {
+			return errors.Wrap(err, "fetching ToggleButton ConfigMap")
+		}
+
+		var desiredValue string
+		if newOnValue {
+			desiredValue = ss.OnValue
+		} else {
+			desiredValue = ss.OffValue
+		}
+
+		if cm.Data == nil {
+			cm.Data = make(map[string]string)
+		}
+
+		currentValue, ok := cm.Data[ss.Key]
+
+		if !ok || currentValue != desiredValue {
+			cm.Data[ss.Key] = desiredValue
+			err := r.ctrlClient.Update(ctx, &cm)
+			if err != nil {
+				return errors.Wrap(err, "updating ConfigMap with ToggleButton value")
+			}
+		}
+
+		r.lastClickProcessTimes[tb.Name] = time.Now()
+	}
+
+	return nil
+}
+
+func (r *Reconciler) manageToggleButton(ctx context.Context, tb *v1alpha1.ToggleButton) error {
+	ss := tb.Spec.StateSource.ConfigMap
+	var cm v1alpha1.ConfigMap
+	err := r.ctrlClient.Get(ctx, types.NamespacedName{Name: ss.Name}, &cm)
+	// TODO(matt) what should we do here?
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return errors.Wrapf(err, "fetching ToggleButton %q ConfigMap %q", tb.Name, ss.Name)
+	}
+
+	if cm.Data != nil {
+		var newOn bool
+		cmVal, ok := cm.Data[ss.Key]
+		if ok {
+			switch cmVal {
+			case ss.OnValue:
+				newOn = true
+			case ss.OffValue:
+				newOn = false
+			default:
+				newOn = tb.Spec.DefaultOn
+			}
+		} else {
+			newOn = tb.Spec.DefaultOn
+		}
+
+		if newOn != tb.Status.On {
+			tb.Status.On = newOn
+			tb.Status.LastChange = metav1.NowMicro()
```

Comment:
i think this is going to have race conditions and idempotency problems.

i'm not sure why this is part of Status at all though? i think these libraries generally try to be very level-based.